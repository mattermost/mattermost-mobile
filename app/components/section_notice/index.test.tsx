// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {type ComponentProps} from 'react';

import {fireEvent, renderWithEverything} from '@test/intl-test-helper';
import TestHelper from '@test/test_helper';

import SectionNotice from '.';

import type Database from '@nozbe/watermelondb/Database';

function getBaseProps(): ComponentProps<typeof SectionNotice> {
    return {
        title: 'Some title',
        text: 'Some text',
        type: 'info',
        isDismissable: true,
        primaryButton: {
            onClick: jest.fn(),
            text: 'primary button',
            leadingIcon: 'chevron-left',
            trailingIcon: 'chevron-right',
            loading: true,
        },
        secondaryButton: {
            onClick: jest.fn(),
            text: 'secondary button',
            leadingIcon: 'chevron-left',
            trailingIcon: 'chevron-right',
            loading: true,
        },
        linkButton: {
            onClick: jest.fn(),
            text: 'link button',
            leadingIcon: 'chevron-left',
            trailingIcon: 'chevron-right',
            loading: true,
        },
        onDismissClick: jest.fn(),
    };
}

describe('Section notice', () => {
    let database: Database;

    beforeAll(async () => {
        const server = await TestHelper.setupServerDatabase();
        database = server.database;
    });

    it('match snapshot', () => {
        const props = getBaseProps();
        const wrapper = renderWithEverything(<SectionNotice {...props}/>, {database});
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should show buttons only if defined', () => {
        const props = getBaseProps();
        const wrapper = renderWithEverything(<SectionNotice {...props}/>, {database});

        const primaryText = props.primaryButton!.text;
        const secondaryText = props.secondaryButton!.text;
        const linkText = props.linkButton!.text;

        expect(wrapper.getAllByRole('button')).toHaveLength(4);
        expect(wrapper.getByText(primaryText)).toBeVisible();
        expect(wrapper.getByText(secondaryText)).toBeVisible();
        expect(wrapper.getByText(linkText)).toBeVisible();
        expect(wrapper.getByTestId('sectionNoticeDismissButton')).toBeVisible();

        props.primaryButton = undefined;
        wrapper.rerender(<SectionNotice {...props}/>);
        expect(wrapper.getAllByRole('button')).toHaveLength(3);
        expect(wrapper.queryByText(primaryText)).not.toBeVisible();
        expect(wrapper.getByText(secondaryText)).toBeVisible();
        expect(wrapper.getByText(linkText)).toBeVisible();
        expect(wrapper.getByTestId('sectionNoticeDismissButton')).toBeVisible();

        props.secondaryButton = undefined;
        wrapper.rerender(<SectionNotice {...props}/>);
        expect(wrapper.getAllByRole('button')).toHaveLength(2);
        expect(wrapper.queryByText(primaryText)).not.toBeVisible();
        expect(wrapper.queryByText(secondaryText)).not.toBeVisible();
        expect(wrapper.getByText(linkText)).toBeVisible();
        expect(wrapper.getByTestId('sectionNoticeDismissButton')).toBeVisible();

        props.linkButton = undefined;
        wrapper.rerender(<SectionNotice {...props}/>);
        expect(wrapper.getAllByRole('button')).toHaveLength(1);
        expect(wrapper.queryByText(primaryText)).not.toBeVisible();
        expect(wrapper.queryByText(secondaryText)).not.toBeVisible();
        expect(wrapper.queryByText(linkText)).not.toBeVisible();
        expect(wrapper.getByTestId('sectionNoticeDismissButton')).toBeVisible();

        props.isDismissable = false;
        wrapper.rerender(<SectionNotice {...props}/>);
        expect(wrapper.queryAllByRole('button')).toHaveLength(0);
        expect(wrapper.queryByText(primaryText)).not.toBeVisible();
        expect(wrapper.queryByText(secondaryText)).not.toBeVisible();
        expect(wrapper.queryByText(linkText)).not.toBeVisible();
        expect(wrapper.queryByTestId('sectionNoticeDismissButton')).not.toBeVisible();
    });

    it('should show the correct icon on each section type', () => {
        const props = getBaseProps();

        props.type = 'info';
        const wrapper = renderWithEverything(<SectionNotice {...props}/>, {database});
        let icon = wrapper.getByTestId('sectionNoticeHeaderIcon');
        expect(icon).toBeVisible();
        expect(icon.props).toHaveProperty('name', 'information-outline');

        props.type = 'danger';
        wrapper.rerender(<SectionNotice {...props}/>);
        icon = wrapper.getByTestId('sectionNoticeHeaderIcon');
        expect(icon).toBeVisible();
        expect(icon.props).toHaveProperty('name', 'alert-outline');

        props.type = 'hint';
        wrapper.rerender(<SectionNotice {...props}/>);
        icon = wrapper.getByTestId('sectionNoticeHeaderIcon');
        expect(icon).toBeVisible();
        expect(icon.props).toHaveProperty('name', 'lightbulb-outline');

        props.type = 'success';
        wrapper.rerender(<SectionNotice {...props}/>);
        icon = wrapper.getByTestId('sectionNoticeHeaderIcon');
        expect(icon).toBeVisible();
        expect(icon.props).toHaveProperty('name', 'check');

        props.type = 'warning';
        wrapper.rerender(<SectionNotice {...props}/>);
        icon = wrapper.getByTestId('sectionNoticeHeaderIcon');
        expect(icon).toBeVisible();
        expect(icon.props).toHaveProperty('name', 'alert-outline');

        props.type = 'welcome';
        wrapper.rerender(<SectionNotice {...props}/>);
        expect(wrapper.queryByTestId('sectionNoticeHeaderIcon')).not.toBeVisible();
    });

    it('should have the correct background on each section type', () => {
        const props = getBaseProps();

        props.type = 'info';
        const wrapper = renderWithEverything(<SectionNotice {...props}/>, {database});
        let container = wrapper.getByTestId('sectionNoticeContainer');
        expect(container).toHaveStyle({backgroundColor: 'rgba(93,137,234,0.08)'});

        props.type = 'danger';
        wrapper.rerender(<SectionNotice {...props}/>);
        container = wrapper.getByTestId('sectionNoticeContainer');
        expect(container).toHaveStyle({backgroundColor: 'rgba(210,75,78,0.08)'});

        props.type = 'hint';
        wrapper.rerender(<SectionNotice {...props}/>);
        container = wrapper.getByTestId('sectionNoticeContainer');
        expect(container).toHaveStyle({backgroundColor: 'rgba(93,137,234,0.08)'});

        props.type = 'success';
        wrapper.rerender(<SectionNotice {...props}/>);
        container = wrapper.getByTestId('sectionNoticeContainer');
        expect(container).toHaveStyle({backgroundColor: 'rgba(61,184,135,0.08)'});

        props.type = 'warning';
        wrapper.rerender(<SectionNotice {...props}/>);
        container = wrapper.getByTestId('sectionNoticeContainer');
        expect(container).toHaveStyle({backgroundColor: 'rgba(255,188,31,0.08)'});

        props.type = 'welcome';
        wrapper.rerender(<SectionNotice {...props}/>);
        container = wrapper.getByTestId('sectionNoticeContainer');
        expect(container).toHaveStyle({backgroundColor: 'rgba(63,67,80,0.04)'});
    });

    it('all buttons perform the expected action', () => {
        const props = getBaseProps();
        const wrapper = renderWithEverything(<SectionNotice {...props}/>, {database});

        fireEvent.press(wrapper.getByText(props.primaryButton!.text));
        expect(props.primaryButton!.onClick).toHaveBeenCalled();
        expect(props.secondaryButton!.onClick).not.toHaveBeenCalled();
        expect(props.linkButton!.onClick).not.toHaveBeenCalled();
        expect(props.onDismissClick!).not.toHaveBeenCalled();

        jest.clearAllMocks();

        fireEvent.press(wrapper.getByText(props.secondaryButton!.text));
        expect(props.primaryButton!.onClick).not.toHaveBeenCalled();
        expect(props.secondaryButton!.onClick).toHaveBeenCalled();
        expect(props.linkButton!.onClick).not.toHaveBeenCalled();
        expect(props.onDismissClick!).not.toHaveBeenCalled();

        jest.clearAllMocks();

        fireEvent.press(wrapper.getByText(props.linkButton!.text));
        expect(props.primaryButton!.onClick).not.toHaveBeenCalled();
        expect(props.secondaryButton!.onClick).not.toHaveBeenCalled();
        expect(props.linkButton!.onClick).toHaveBeenCalled();
        expect(props.onDismissClick!).not.toHaveBeenCalled();

        jest.clearAllMocks();

        fireEvent.press(wrapper.getByTestId('sectionNoticeDismissButton'));
        expect(props.primaryButton!.onClick).not.toHaveBeenCalled();
        expect(props.secondaryButton!.onClick).not.toHaveBeenCalled();
        expect(props.linkButton!.onClick).not.toHaveBeenCalled();
        expect(props.onDismissClick!).toHaveBeenCalled();
    });
});
