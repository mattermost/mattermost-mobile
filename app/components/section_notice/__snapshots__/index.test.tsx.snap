// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section notice match snapshot 1`] = `
<View
  style={
    [
      {
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 1,
      },
      {
        "backgroundColor": "rgba(93,137,234,0.08)",
        "borderColor": "rgba(93,137,234,0.16)",
      },
    ]
  }
  testID="sectionNoticeContainer"
>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "gap": 12,
        "padding": 16,
      }
    }
  >
    <Icon
      name="information-outline"
      size={20}
      style={
        {
          "color": "#5d89ea",
        }
      }
      testID="sectionNoticeHeaderIcon"
    />
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "column",
          "gap": 8,
        }
      }
    >
      <Text
        style={
          {
            "color": "#3f4350",
            "fontFamily": "OpenSans-SemiBold",
            "fontSize": 16,
            "fontWeight": "600",
            "lineHeight": 24,
            "margin": 0,
          }
        }
      >
        Some title
      </Text>
      <View
        style={
          [
            {
              "alignItems": "flex-start",
              "flexDirection": "row",
              "flexWrap": "wrap",
            },
          ]
        }
        testID="markdown_paragraph"
      >
        <Text>
          <Text
            selectable={false}
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              }
            }
            testID="markdown_text"
          >
            Some text
          </Text>
        </Text>
      </View>
      <View
        style={
          {
            "gap": 8,
            "marginTop": 8,
          }
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 2,
              },
              undefined,
              false,
            ]
          }
          testID="RNE_BUTTON_WRAPPER"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": false,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#1c58d9",
                  "borderColor": "#2089dc",
                  "borderRadius": 4,
                  "borderWidth": 0,
                  "flex": 0,
                  "flexDirection": "row",
                  "height": 40,
                  "justifyContent": "center",
                  "padding": 8,
                  "paddingHorizontal": 20,
                  "paddingVertical": 12,
                }
              }
            >
              <View
                style={
                  [
                    {
                      "flexDirection": "row",
                    },
                    {
                      "minHeight": 18,
                    },
                  ]
                }
              >
                <Icon
                  color="#ffffff"
                  name="chevron-right"
                  size={18}
                  style={
                    {
                      "marginRight": 7,
                    }
                  }
                />
                <Text
                  numberOfLines={1}
                  style={
                    [
                      [
                        {
                          "alignItems": "center",
                          "fontFamily": "OpenSans-SemiBold",
                          "fontWeight": "600",
                          "justifyContent": "center",
                          "padding": 1,
                          "textAlignVertical": "center",
                        },
                        {
                          "fontSize": 14,
                          "lineHeight": 14,
                          "marginTop": 3,
                        },
                        {
                          "color": "#ffffff",
                        },
                      ],
                      undefined,
                    ]
                  }
                >
                  primary button
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 2,
              },
              undefined,
              false,
            ]
          }
          testID="RNE_BUTTON_WRAPPER"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": false,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "rgba(28,88,217,0.08)",
                  "borderColor": "#2089dc",
                  "borderRadius": 4,
                  "borderWidth": 0,
                  "flex": 0,
                  "flexDirection": "row",
                  "height": 40,
                  "justifyContent": "center",
                  "padding": 8,
                  "paddingHorizontal": 20,
                  "paddingVertical": 12,
                }
              }
            >
              <View
                style={
                  [
                    {
                      "flexDirection": "row",
                    },
                    {
                      "minHeight": 18,
                    },
                  ]
                }
              >
                <Icon
                  color="#1c58d9"
                  name="chevron-right"
                  size={18}
                  style={
                    {
                      "marginRight": 7,
                    }
                  }
                />
                <Text
                  numberOfLines={1}
                  style={
                    [
                      [
                        {
                          "alignItems": "center",
                          "fontFamily": "OpenSans-SemiBold",
                          "fontWeight": "600",
                          "justifyContent": "center",
                          "padding": 1,
                          "textAlignVertical": "center",
                        },
                        {
                          "fontSize": 14,
                          "lineHeight": 14,
                          "marginTop": 3,
                        },
                        {
                          "color": "#1c58d9",
                        },
                      ],
                      undefined,
                    ]
                  }
                >
                  secondary button
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 2,
              },
              undefined,
              false,
            ]
          }
          testID="RNE_BUTTON_WRAPPER"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": false,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 4,
                  "borderWidth": 0,
                  "flex": 0,
                  "flexDirection": "row",
                  "height": 40,
                  "justifyContent": "center",
                  "padding": 8,
                  "paddingHorizontal": 20,
                  "paddingVertical": 12,
                }
              }
            >
              <View
                style={
                  [
                    {
                      "flexDirection": "row",
                    },
                    {
                      "minHeight": 18,
                    },
                  ]
                }
              >
                <Icon
                  color="#1c58d9"
                  name="chevron-right"
                  size={18}
                  style={
                    {
                      "marginRight": 7,
                    }
                  }
                />
                <Text
                  numberOfLines={1}
                  style={
                    [
                      [
                        {
                          "alignItems": "center",
                          "fontFamily": "OpenSans-SemiBold",
                          "fontWeight": "600",
                          "justifyContent": "center",
                          "padding": 1,
                          "textAlignVertical": "center",
                        },
                        {
                          "fontSize": 14,
                          "lineHeight": 14,
                          "marginTop": 3,
                        },
                        {
                          "color": "#1c58d9",
                        },
                      ],
                      undefined,
                    ]
                  }
                >
                  link button
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="button"
    style={
      {
        "alignItems": "center",
        "height": 32,
        "justifyContent": "center",
        "position": "absolute",
        "right": 10,
        "top": 10,
        "width": 32,
      }
    }
    testID="sectionNoticeDismissButton"
  >
    <Icon
      name="close"
      size={18}
    />
  </View>
</View>
`;
